### PR TITLE
DAOS-8039 control: Add retries for telemetry queries

### PR DESCRIPTION
**Pull request by kjacque**
Add retries to http endpoint requests, and clarify the error
returned when the request times out. This applies to both metrics
list and metrics query.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Review by mjmac — COMMENTED** [review]
Looks good, just a couple of minor things to address.

File: src/control/lib/control/http.go [hunk 1]
```diff
@@ -20,13 +21,101 @@ import (
...
+func (r *httpReq) canRetry(err error, cur uint) bool {
+	if r == nil || r.url == nil {
+		return false
+	}
+
+	if cur >= httpMaxRetries {
+		return false
+	}
+
+	if err.Error() == HTTPReqTimedOut(r.getURL().String()).Error() {
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/lib/control/http.go]: Good idea. Fixed.

File: src/control/lib/control/http.go [hunk 2]
```diff
@@ -57,6 +146,9 @@ func httpGetBody(ctx context.Context, url *url.URL, get httpGetFn) ([]byte, erro
 
 	select {
 	case <-httpCtx.Done():
+		if errors.Is(httpCtx.Err(), context.DeadlineExceeded) {
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/lib/control/http.go]: Ah, I was thinking we were on 1.13+. Fixed.

**Review by mjmac — APPROVED** [review]

File: src/control/lib/control/http.go [hunk 3]
```diff
@@ -20,13 +21,101 @@ import (
...
+		if err == nil {
+			break
+		}
+
+		if !req.canRetry(err, i) {
+			return nil, err
+		}
+
+		time.Sleep(req.retryAfter(0))
+		if err = req.onRetry(ctx, i); err != nil {
```
Comment (dpquigl) [inline src/control/lib/control/http.go:108]: What is the purpose of this call since it looks like onRetry always returns nil?

**Review by dpquigl — APPROVED** [review]

**Pull request merged**